### PR TITLE
Relative Actions

### PR DESCRIPTION
**Pull request by donaldchen34**
Added undo function and routes for undo and clear

Issue: Closes #

### Frontend:
- 
### Backend:
-

### CODE REVIEW

**Review by joshcwinton — APPROVED** [review]
Looks good 👍 

I think the ML stuff got pulled in here because the branch isn't based on main but nbd

LGTM 💰

File: backend/app.py [hunk 1]
```diff
@@ -88,6 +88,20 @@ def history():
...
 
 
+@app.route('/undo', methods=['GET'])
+def undo_route():
+    if request.method == 'GET':
+        undo()
+    return None
+
+
+@app.route('/clear', methods=['GET'])
```
Comment (JasperCheung) [inline backend/app.py]: should return a message like "Undo Action!"
 ```def chatbot_route():
    if request.method == 'POST':
        post_data = request.get_json()
        user_res = post_data["user"]
        bot_res = chatbot(user_res)
        return jsonify({"SHRDLU": bot_res})
    return jsonify({"get": "requested"})
```

File: backend/app.py [hunk 2]
```diff
@@ -88,6 +88,20 @@ def history():
...
+def clear_route():
```
Comment (JasperCheung) [inline backend/app.py:100]: same thing as above

**Review by JasperCheung — CHANGES_REQUESTED** [review]

**Review by svhnstorage — APPROVED** [review]

File: backend/environment.py [hunk 3]
```diff
@@ -74,11 +72,32 @@ def holdShape(row, col):
...
 
 def clearBoard():
-    global GRID
+    global GRID, HISTORY, MESSAGES
+
     for x in range(GRID_SIZE):
         for y in range(GRID_SIZE):
             GRID[x][y] = ("", "", 0)
 
+    MESSAGES.append({"name": "Me", "Clear board": inputMessage})
```
Comment (joshcwinton) [inline backend/environment.py:81]: the message part of these should have the key `text` and value the actual message

i.e. `{"name": "Me", "text": inputMessage}`

see `updateHistory()`

**Review by joshcwinton — CHANGES_REQUESTED** [review]

**Comment by JasperCheung** [comment]
Refer to newer pr

**Pull request closed without merging**